### DISCUSSION TITLE
Python 3.10 compatibility

### DISCUSSION THREAD

**Discussion by fbarfi** [alan-turing-institute/sktime#2651] [Q&A]
Can sktime be compatible with python3.10.4? thanks.

**Comment by fkiraly** [comment]
Relevant issue here: https://github.com/alan-turing-institute/sktime/issues/2079

@lmmentel, @freddyaboulton, what´s the newest? Anything we need help on?

**Reply by lmmentel** [reply]
Sktime is largely compatible, but there are soft dependencies that are not yet compatible with python 3.10. This is at the moment preventing us from releasing a compatible version.

**Reply by fkiraly** [reply]
hmm, why would that prevent us?

We can simply raise warnings and errors with a descriptive message regarding python compatibility, in those problematic soft dependencies? Might take a little bit of tinkering with the soft dependency utility, but surely soft dependencies should not prevent us from supporting python 3.10? That´s also not sustainable in the limit of infinite soft dependencies. I think we should generally decouple python support of sktime from that of soft dependencies.

**Reply by lmmentel** [reply]
Practically, when running tests we run test with all soft deps which means test will fail for modules that depend on incompatible deps. We can either release with broken tests/not running tests or not release at all until either we can run tests without soft deps or soft deps support python 3.10.

**Reply by fkiraly** [reply]
I'm not sure whether we are talking across purposes.

I am making a third suggestion not covered by your either-or, that seems like a false dichotomy to me based either on the assumption that we must support all soft dependencies at 3.10, or on the assumption that we must run all tests, even if some estimators do not support the python version. Why would we need to do either, it is not our reponsibility to keep the other package up to date.

The third proposal I am referring to combines the following:

* useful, descriptive user warnings when trying to load or instantiate an estimator that is not supported. E.g., "this estimator requires python 3.9.X or lower".
* removing affected estimators from tests on versions where they are not supported

**Reply by fkiraly** [reply]
here, a proof of concept: https://github.com/alan-turing-institute/sktime/pull/2661

**Comment by fkiraly** [comment]
From issue #2715, discussion between @fkiraly and @2-5

@2-5 asking:
On a more general note, why forcefully limit future version support?

@fkiraly:
If not limiting it, automatic version resolution might automatically install python 3.10?
Which is not fully supported, and hence may produce errors in a "default install", to avoid this, 3.9 or lower would now have to be expressly requested.

@2-5:
> If not limiting it, automatic version resolution might automatically install python 3.10? 

I guess you mean when you do `conda install sktime` it might install Python 3.10? I never used `conda`, I wasn't aware it can install Python too, so I guess you are right.

`pip` however will never install or upgrade Python, that's not something it can do.

I did a little googling on best practices, and I found this (long) blog post on this issue from Henry Schreiner, Python Packaging Authority member:

https://iscinumpy.dev/post/bound-version-constraints/

> Bound version constraints (upper caps) are starting to show up in the Python ecosystem. This is causing real world problems with libraries following this recommendation, and is likely to continue to get worse; this practice does not scale to large numbers of libraries or large numbers of users. In this discussion I would like to explain why always providing an upper limit causes far more harm than good even for true SemVer libraries, why libraries that pin upper limits require more frequent updates rather than less, and why it is not scalable. After reading this, hopefully you will always consider every cap you add, you will know the (few) places where pinning an upper limit is reasonable, and will possibly even avoid using libraries that pin upper limits needlessly until the author updates them to remove these pins.

> ...

> **Never provide an upper cap to your Python version. I generally will not use a library that has an upper cap to the Python version**

> ...

> TL;DR
> Capping dependencies has long term negative effects, especially for libraries, and should never be taken lightly. A library is not installed in isolation; it has to live with other libraries in a shared environment. Only add a cap if a dependency is known to be incompatible or there is a high (>75%) chance of it being incompatible in its next release. Do not cap by default - capping dependencies makes your software incompatible with other libraries that also have strict lower limits on dependencies, and limits future fixes. Anyone can fix a missing cap, but users cannot fix an over restrictive cap causing solver errors. It also encourages hiding issues until they become harder to fix, it does not scale to larger systems, it limits your ability to access security and bugfix updates, and some tools (Poetry) force these bad decisions on your downstream users if you make them. **Never cap Python, it is fundamentally broken at the moment.** Also, even packing capping has negative consequences that can produce unexpected solves.
> ...
> If you need a TL;DR for the TL;DR, I’ll just [quote](https://snarky.ca/why-i-dont-like-semver/) Python Steering Council Member and packaging expert Brett Cannon:
>> Libraries/packages should be setting a floor, and if necessary excluding known buggy versions, but otherwise don’t cap the maximum version as you can’t predict future compatibility

**Reply by fkiraly** [reply]
@2-5, thanks, interesting expert opinion!

Now, if we don't want to upper bound python, how would it practically work with the dependencies that *do* have upper compatibility bounds that impact functionality?
